### ISSUE TITLE
Cannot index into a null array when creating a container from image businesscentral/onprem:16.0.11240.12085-gb

### ISSUE THREAD

**Issue by darrenwaller** [microsoft/navcontainerhelper#1172]
**Describe the issue**
The following error is now being seen when we create a container for onprem 16.0.11240.12085-gb

WARNING: Waiting for service 'World Wide Web Publishing Service (W3SVC)' to start...
Cannot index into a null array.
at <ScriptBlock>, C:\Run\navstart.ps1: line 100
at <ScriptBlock>, C:\Run\start.ps1: line 327
at <ScriptBlock>, <No file>: line 1**Scripts used to create container and cause the issue**
```

I have only started seeing this since generic image was updated today

mcr.microsoft.com/dynamicsnav   10.0.14393.3808-generic   2c605cf355db        20 hours ago        16.5GB

I deleted artifacts cache and deleted the image businesscentral/onprem:16.0.11240.12085-gb then tried to run again but still get the same problem.

We are running

New-BCContainer -accept_eula -imageName $imageName -artifactUrl $artifactUrl -containerName $containerName -licenseFile $License.FullName `
            -auth $auth -updateHosts -accept_outdated -myScripts $myScripts -locale $Locale -TimeZoneId $TimeZoneId `
            -additionalParameters $additionalParameters -databaseServer $applicationDatabaseServer -databaseInstance $applicationDatabaseInstance -databaseName $applicationDatabaseName `
            -databaseCredential $dbCredential `
            -WebClientPort $WebClientPort  -ClientServicesPort $ClientServicesPort -SoapServicesPort $SoapServicesPort -ODataServicesPort $ODataServicesPort -DeveloperServicesPort $DeveloperServicesPort -ManagementServicesPort $ManagementServicesPort `
            -PublishPorts $PublishPorts

with these values
$auth = "Windows"
$artifactType = "onprem"                                                                                     
$artifactVersion = "16.0.11240.12085"                                                                        
$artifactCountry = "gb"                                                                                      
$containerName = "TESTCONTAINER16"
$encryptionSecurePassword = (ConvertTo-SecureString "xxxxxxxxxxxxx" -AsPlainText -Force)                          
$dbCredential = New-Object pscredential "docker", (ConvertTo-SecureString "xxxxxxxxxxxxxx" -AsPlainText -Force) 
$applicationDatabaseServer = "mbssql.ad.mydomain.co.uk"                                               
$applicationDatabaseInstance = "SQL2017"                                                                     
$applicationDatabaseName = "TESTDB16"                                                                    
$additionalParameters = @("--security-opt credentialspec=file://blue_testcontainerV16.json")
-WebClientPort 8280 `
-ManagementServicesPort 7245 `
-ClientServicesPort 7246 `
-SoapServicesPort 7247 `
 -ODataServicesPort 7248 `
-DeveloperServicesPort 7249 `
-PublishPorts @("8280", "8281", "7245", "7246", "7247", "7248", "7249")

SQL is installed on another server in the domain and we are using gMSA.

I have just created a BC V14 container and that was fine.

Puzzled why V16 will not work and it fails on line 100 in navstart.ps1 (Line shown below)

$WebClientFolder = (Get-Item "C:\Program Files\Microsoft Dynamics NAV\*\Web Client")[0]
```

**Full output of scripts**
```
NavContainerHelper is version 0.7.0.12
NavContainerHelper is running as administrator
Host is Microsoft Windows Server 2016 Standard Version 10.0.14393 Build 14393
Docker Client -Enterprise Version is 19.03.5
Docker Server -Enterprise Version is 19.03.5
...
```

**Screenshots**

**Additional context**
- does it happen all the time Yes
- did it use to work Yes before the generic windows image was updated today.

**Comment by freddydk** [comment]
I would really like the full output - I have a lot of debug info in that, which probably could point me to the reason for this.
Also, could you try to add "--env doNotUseNewFolder=Y" to your additionalparameters and retry?

**Comment by darrenwaller** [comment]
Hi Freddy,

Please see below output from using doNotUseNewFolder and without. Interestingly if I create a container without using a database on another server and without using credentialspec and publishing ports with database auth etc.. it builds a new image (Not sure why it did this when the image existed) then the container creates fine. If I then use my original script it also creates the container fine. I only see the problem if the image is first created when I use my original script. Anyway I hope you can find something I am missing.

doNotUseNewFolder gives me a different error see below

Modifying Service Tier Config File with Instance Specific Settings
Starting Service Tier
Registering event sources
Creating DotNetCore Web Server Instance
The running command stopped because the preference variable "ErrorActionPreference" or common parameter is set to Stop: The term ' ' is not recognized as the name of a cmdlet, function, script file, or
 operable program. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
at <ScriptBlock>, C:\Program Files\Microsoft Dynamics NAV\160\Web Client\Modules\NAVWebClientManagement\NAVWebClientManagement.psm1: line 1
at <ScriptBlock>, C:\Run\SetupWebClient.ps1: line 25
at <ScriptBlock>, C:\Run\navstart.ps1: line 157
at <ScriptBlock>, C:\Run\start.ps1: line 327
at <ScriptBlock>, <No file>: line 1

Full output below without using doNotUseNewFolder.

FULL OUTPUT:
NavContainerHelper is version 0.7.0.21
NavContainerHelper is running as administrator
Host is Microsoft Windows Server 2016 Standard - ltsc2016
Docker Client Version is 19.03.5
Docker Server Version is 19.03.5
ArtifactUrl and ImageName specified
Fetching all docker images
Image businesscentral/onprem:16.0.11240.12085-gb doesn't exist
Building image businesscentral/onprem:16.0.11240.12085-gb based on https://bcartifacts.azureedge.net/onprem/16.0.11240.12085/gb
Pulling latest image mcr.microsoft.com/dynamicsnav:10.0.14393.3808-generic
10.0.14393.3808-generic: Pulling from dynamicsnav
Generic Tag: 0.1.0.12
Container OS Version: 10.0.14393.3808 (ltsc2016)
Host OS Version: 10.0.14393.3808 (ltsc2016)
Copying Platform Artifacts
Copying Installers
Copying ConfigurationPackages
Copying Applications
c:\bcartifacts.cache\tmp637316447175575706
Sending build context to Docker daemon  1.073GB

Step 1/6 : FROM mcr.microsoft.com/dynamicsnav:10.0.14393.3808-generic
 ---> 2c605cf355db
Step 2/6 : ENV DatabaseServer=localhost DatabaseInstance=SQLEXPRESS DatabaseName=CRONUS IsBcSandbox=N artifactUrl=https://bcartifacts.azureedge.net/onprem/16.0.11240.12085/gb
 ---> Running in 43016d867278
Removing intermediate container 43016d867278
 ---> 12d1cffe8d00
Step 3/6 : COPY my /run/
 ---> bee518a228ee
Step 4/6 : COPY NAVDVD /NAVDVD/
 ---> 2fd316f2ea91
Step 5/6 : RUN \Run\start.ps1 -installOnly
 ---> Running in 904b588bc59e
Using installer from C:\Run\150-new
Installing Business Central
Installing from DVD
Starting Local SQL Server
WARNING: Waiting for service 'SQL Server (SQLEXPRESS) (MSSQL$SQLEXPRESS)' to 
start...
WARNING: Waiting for service 'SQL Server (SQLEXPRESS) (MSSQL$SQLEXPRESS)' to 
start...
WARNING: Waiting for service 'SQL Server (SQLEXPRESS) (MSSQL$SQLEXPRESS)' to 
start...
WARNING: Waiting for service 'SQL Server (SQLEXPRESS) (MSSQL$SQLEXPRESS)' to 
start...
WARNING: Waiting for service 'SQL Server (SQLEXPRESS) (MSSQL$SQLEXPRESS)' to 
start...
WARNING: Waiting for service 'SQL Server (SQLEXPRESS) (MSSQL$SQLEXPRESS)' to 
start...
WARNING: Waiting for service 'SQL Server (SQLEXPRESS) (MSSQL$SQLEXPRESS)' to 
start...
WARNING: Waiting for service 'SQL Server (SQLEXPRESS) (MSSQL$SQLEXPRESS)' to 
start...
WARNING: Waiting for service 'SQL Server (SQLEXPRESS) (MSSQL$SQLEXPRESS)' to 
start...
WARNING: Waiting for service 'SQL Server (SQLEXPRESS) (MSSQL$SQLEXPRESS)' to 
start...
Starting Internet Information Server
Copying Service Tier Files
Copying PowerShell Scripts
Copying dependencies
Copying ReportBuilder
Importing PowerShell Modules
Skipping restore of Cronus database
Modifying Business Central Service Tier Config File for Docker
Creating Business Central Service Tier
Installing SIP crypto provider: 'C:\Windows\System32\NavSip.dll'
Installation took 117 seconds
Installation complete
Removing intermediate container 904b588bc59e
 ---> 5c01a774013f
Step 6/6 : LABEL legal="http://go.microsoft.com/fwlink/?LinkId=837447"       created="202007291742"       nav="2004"       cu="rtm"       skipdatabase="Y" country="gb"       version="16.0.11240.12085" 
      platform="16.0.11233.12078"
 ---> Running in 70aba83d5b89
Removing intermediate container 70aba83d5b89
 ---> 9c8a4642acdf
Successfully built 9c8a4642acdf
Successfully tagged businesscentral/onprem:16.0.11240.12085-gb
Building image took 937 seconds
Using image businesscentral/onprem:16.0.11240.12085-gb
Publishing port 8280
Publishing port 8281
Publishing port 7245
Publishing port 7246
Publishing port 7247
Publishing port 7248
Publishing port 7249
Creating Container DEMOV16
Version: 16.0.11240.12085-gb
Style: onprem
Platform: 16.0.11233.12078
Generic Tag: 0.1.0.12
Container OS Version: 10.0.14393.3808 (ltsc2016)
Host OS Version: 10.0.14393.3808 (ltsc2016)
Using locale en-GB
Using process isolation
Disabling the standard eventlog dump to container log every 2 seconds (use -dumpEventLog to enable)
Using license file C:\ContainerInstalls\DEMOV16\License\MyLicense.flf
Additional Parameters:
--security-opt credentialspec=file://blue_DemoV16.json
Files in C:\ProgramData\NavContainerHelper\Extensions\DEMOV16\my:
- Add-Ins
- AdditionalOutput.ps1
- license.flf
- MainLoop.ps1
- SetupVariables.ps1
- updatehosts.ps1
Creating container DEMOV16 from image businesscentral/onprem:16.0.11240.12085-gb
bf756ad99534a8885cc9bdea67c13586857a3f2cf567f3b12a620edfd867ab76
Waiting for container DEMOV16 to be ready
Initializing...
Setting host.containerhelper.internal to 172.29.32.1 in container hosts file
Starting Container
Hostname is DEMOV16
PublicDnsName is DEMOV16
Using Windows Authentication
Starting Internet Information Server
Cannot index into a null array.
at <ScriptBlock>, C:\Run\navstart.ps1: line 100
at <ScriptBlock>, C:\Run\start.ps1: line 327
at <ScriptBlock>, <No file>: line 1Error
Initializing...
Setting host.containerhelper.internal to 172.29.32.1 in container hosts file
Starting Container
Hostname is DEMOV16
PublicDnsName is DEMOV16
Using Windows Authentication
Starting Internet Information Server
Cannot index into a null array.
at <ScriptBlock>, C:\Run\navstart.ps1: line 100
at <ScriptBlock>, C:\Run\start.ps1: line 327
at <ScriptBlock>, <No file>: line 1
Initialization of container DEMOV16 failed
At C:\Program Files\WindowsPowerShell\Modules\navcontainerhelper\0.7.0.21\ContainerHandling\Wait-NavContainerReady.ps1:44 char:17
+ ...             throw "Initialization of container $containerName failed" ...
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (Initialization ...LAYERV16 failed:String) [], RuntimeException
    + FullyQualifiedErrorId : Initialization of container BLUELAYERV16 failed

**Comment by freddydk** [comment]
If you add 

-useGenericImage "$(Get-BestGenericImageName)-0.1.0.7"

Then it works?

**Comment by darrenwaller** [comment]
Hi Freddy,

Works fine using -useGenericImage "$(Get-BestGenericImageName)-0.1.0.7"

**Comment by freddydk** [comment]
I have created a Windows Server 2016 with the same build and I cannot repro this.
Does this happen with other artifacts as well?
If you use Flush-ContainerHelperCache - does it work then?
Does the host machine have sufficient free space on the hard drive?

**Comment by freddydk** [comment]
What are you using as imagename? - this "businesscentral/onprem:16.0.11240.12085-gb"?
Normally you should just use "myimg" or something like that - then the containerhelper itself will add a tag indicating the version number etc. 
Don't think this is the problem though

**Comment by freddydk** [comment]
If you look in this folder: C:\bcartifacts.cache\onprem\16.0.11240.12085\platform on your machine.
Does it look like this?
![image](https://user-images.githubusercontent.com/10775043/88899449-5abf7400-d24e-11ea-9bfe-2d463f4d75d2.png)

**Comment by freddydk** [comment]
I can repro the error if I delete the WebClient folder from the artifacts.
I have to assume that the artifacts have been damaged or something went wrong during unpacking.

If you use
Flush-ContainerHelperCache (to remove caches)
Remove-BcContainer to remove containers
docker rmi to remove old images

and then retry - that should work.

**Comment by darrenwaller** [comment]
Hi Freddy, I will try the above now, might take a while

**Comment by darrenwaller** [comment]
I did check the artifacts folder and WebClient did exist

**Comment by freddydk** [comment]
and if you do

dir /s > webclient.txt

In a cmd prompt inside the webclient folder - and attach that?

**Comment by darrenwaller** [comment]
Hi Freddy,

I have isolated the problem. This only seems to happen using  generic image 10.0.14393.3808-generic-0.1.0.12

When the image businesscentral/onprem:16.0.11240.12085-gb is created from the above image and the SQL database is located on another server. For some reason image businesscentral/onprem:16.0.11240.12085-gb gets created ok but any container created from this image does not contain folder "C:\Program Files\Microsoft Dynamics NAV\160\Web Client". This causes the error reported.

Please see script and output below....  (NOTE: If I first create image businesscentral/onprem:16.0.11240.12085-gb where the database is in the container then everything works fine so the problem is using a database located elsewhere when the image businesscentral/onprem:16.0.11240.12085-gb is first created.

SCRIPT:
$artifactType = "onprem"                                                                                     
$artifactVersion = "16.0.11240.12085"                                                                        
$artifactCountry = "gb"                                                                                      
$containerName = "DEMOV16"                                                                              

$dbCredential = New-Object pscredential "docker", (ConvertTo-SecureString "ThisIsMyNewP4ssW0rd!!" -AsPlainText -Force) 
$applicationDatabaseServer = "mysqlserver.ad.mydomain.co.uk"                                               
$applicationDatabaseInstance = "SQL2017"                                                                     
$applicationDatabaseName = "DEMOV16"                                                                    
$additionalParameters = @("--security-opt credentialspec=file://demo_demoV16.json")                 

$IsV16OrAbove = $true
$auth = "Windows"

$artifactUrl = Get-BCArtifactUrl -version $artifactVersion -country $artifactCountry -select Latest -type $artifactType

$imageName = "businesscentral/" + $artifactType + ":" + $artifactVersion + "-" + $artifactCountry

New-BCContainer -accept_eula -imageName $imageName -artifactUrl $artifactUrl -containerName $containerName `
        -auth $auth -updateHosts -accept_outdated -locale $Locale -TimeZoneId $TimeZoneId `
        -additionalParameters $additionalParameters -databaseServer $applicationDatabaseServer -databaseInstance $applicationDatabaseInstance -databaseName $applicationDatabaseName `
        -databaseCredential $dbCredential             
        

OUTPUT:

NavContainerHelper is version 0.7.0.22
NavContainerHelper is running as administrator
Host is Microsoft Windows Server 2016 Standard - ltsc2016
Docker Client Version is 19.03.5
Docker Server Version is 19.03.5
ArtifactUrl and ImageName specified
Fetching all docker images
Image businesscentral/onprem:16.0.11240.12085-gb doesn't exist
Building image businesscentral/onprem:16.0.11240.12085-gb based on https://bcartifacts.azureedge.net/onprem/16.0.11240.12085/gb
Pulling latest image mcr.microsoft.com/dynamicsnav:10.0.14393.3808-generic
10.0.14393.3808-generic: Pulling from dynamicsnav
Generic Tag: 0.1.0.12
Container OS Version: 10.0.14393.3808 (ltsc2016)
Host OS Version: 10.0.14393.3808 (ltsc2016)
Copying Platform Artifacts
Copying Installers
Copying ConfigurationPackages
Copying Applications
c:\bcartifacts.cache\tmp637317109925845004
Sending build context to Docker daemon  1.073GB

Step 1/6 : FROM mcr.microsoft.com/dynamicsnav:10.0.14393.3808-generic
 ---> 2c605cf355db
Step 2/6 : ENV DatabaseServer=localhost DatabaseInstance=SQLEXPRESS DatabaseName=CRONUS IsBcSandbox=N artifactUrl=https://bcartifacts.azureedge.net/onprem/16.0.11240.12085/gb
 ---> Running in 1677318898d7
Removing intermediate container 1677318898d7
 ---> 6a564cd94685
Step 3/6 : COPY my /run/
 ---> 99e100d4afef
Step 4/6 : COPY NAVDVD /NAVDVD/
 ---> be2274ca5e7d
Step 5/6 : RUN \Run\start.ps1 -installOnly
 ---> Running in 36e010b6f83a
Using installer from C:\Run\150-new
Installing Business Central
Installing from DVD
Starting Local SQL Server
WARNING: Waiting for service 'SQL Server (SQLEXPRESS) (MSSQL$SQLEXPRESS)' to 
start...
WARNING: Waiting for service 'SQL Server (SQLEXPRESS) (MSSQL$SQLEXPRESS)' to 
start...
WARNING: Waiting for service 'SQL Server (SQLEXPRESS) (MSSQL$SQLEXPRESS)' to 
start...
WARNING: Waiting for service 'SQL Server (SQLEXPRESS) (MSSQL$SQLEXPRESS)' to 
start...
WARNING: Waiting for service 'SQL Server (SQLEXPRESS) (MSSQL$SQLEXPRESS)' to 
start...
WARNING: Waiting for service 'SQL Server (SQLEXPRESS) (MSSQL$SQLEXPRESS)' to 
start...
WARNING: Waiting for service 'SQL Server (SQLEXPRESS) (MSSQL$SQLEXPRESS)' to 
start...
WARNING: Waiting for service 'SQL Server (SQLEXPRESS) (MSSQL$SQLEXPRESS)' to 
start...
WARNING: Waiting for service 'SQL Server (SQLEXPRESS) (MSSQL$SQLEXPRESS)' to 
start...
WARNING: Waiting for service 'SQL Server (SQLEXPRESS) (MSSQL$SQLEXPRESS)' to 
start...
WARNING: Waiting for service 'SQL Server (SQLEXPRESS) (MSSQL$SQLEXPRESS)' to 
start...
Starting Internet Information Server
Copying Service Tier Files
Copying PowerShell Scripts
Copying dependencies
Copying ReportBuilder
Importing PowerShell Modules
Skipping restore of Cronus database
Modifying Business Central Service Tier Config File for Docker
Creating Business Central Service Tier
Installing SIP crypto provider: 'C:\Windows\System32\NavSip.dll'
Installation took 123 seconds
Installation complete
Removing intermediate container 36e010b6f83a
 ---> 673f179de2eb
Step 6/6 : LABEL legal="http://go.microsoft.com/fwlink/?LinkId=837447"       created="202007301203"       nav="2004"       cu="rtm"       skipdatabase="Y" country="gb"       version="16.0.11240.12085"       platform="16.0.11233.12078"
 ---> Running in 49faac1253c7
Removing intermediate container 49faac1253c7
 ---> 76adb0c5ae94
Successfully built 76adb0c5ae94
Successfully tagged businesscentral/onprem:16.0.11240.12085-gb
Building image took 650 seconds
Using image businesscentral/onprem:16.0.11240.12085-gb
Creating Container DEMOV16
Version: 16.0.11240.12085-gb
Style: onprem
Platform: 16.0.11233.12078
Generic Tag: 0.1.0.12
Container OS Version: 10.0.14393.3808 (ltsc2016)
Host OS Version: 10.0.14393.3808 (ltsc2016)
Using locale en-GB
Using process isolation
Disabling the standard eventlog dump to container log every 2 seconds (use -dumpEventLog to enable)
Using license file C:\BRSContainerInstalls\DEMOV16\License\D365BC Dev License V16 23-07-2020.flf
Additional Parameters:
--security-opt credentialspec=file://demo_DEMOV16.json
Files in C:\ProgramData\NavContainerHelper\Extensions\DEMOV16\my:
- Add-Ins
- AdditionalOutput.ps1
- license.flf
- MainLoop.ps1
- SetupVariables.ps1
- updatehosts.ps1
Creating container DEMOV16 from image businesscentral/onprem:16.0.11240.12085-gb
983b7c838f79b581a64172d2fa3f6be72d60540361c1806a02a19c1f6d6a2ba5
Waiting for container DEMOV16 to be ready
Initializing...
Setting host.containerhelper.internal to 172.29.32.1 in container hosts file
Starting Container
Hostname is DEMOV16
PublicDnsName is DEMOV16
Using Windows Authentication
Starting Internet Information Server
Cannot index into a null array.
at <ScriptBlock>, C:\Run\navstart.ps1: line 100
at <ScriptBlock>, C:\Run\start.ps1: line 327
at <ScriptBlock>, <No file>: line 1Error
Initializing...
Setting host.containerhelper.internal to 172.29.32.1 in container hosts file
Starting Container
Hostname is DEMOV16
PublicDnsName is DEMOV16
Using Windows Authentication
Starting Internet Information Server
Cannot index into a null array.
at <ScriptBlock>, C:\Run\navstart.ps1: line 100
at <ScriptBlock>, C:\Run\start.ps1: line 327
at <ScriptBlock>, <No file>: line 1
Initialization of container DEMOV16 failed
At C:\Program Files\WindowsPowerShell\Modules\navcontainerhelper\0.7.0.22\ContainerHandling\Wait-NavContainerReady.ps1:44 char:17
+ ...             throw "Initialization of container $containerName failed" ...
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (Initialization ...DEMOV16 failed:String) [], RuntimeException
    + FullyQualifiedErrorId : Initialization of container DEMOV16 failed

**Comment by darrenwaller** [comment]
In the output of my previous message (Snippet below). "Copying Web Client Files" does not appear!! 

Starting Internet Information Server
Copying Service Tier Files
Copying PowerShell Scripts
Copying dependencies
Copying ReportBuilder
Importing PowerShell Modules

**Comment by freddydk** [comment]
That was the missing piece - I know what the problem is.
I will create a new generic image asap.
Thanks

**Comment by darrenwaller** [comment]
Great, thank you

**Comment by freddydk** [comment]
Issue was, that the new installer was doing things in parallel to speed up things and ended by doing:

Get-Job | Receive-Job

Problem is, that when it didn't have to restore the database, the other processes were done faster, and the jobs hadn't completed.

This needed to be:

Get-Job | Wait-Job | Receive-Job

**Comment by darrenwaller** [comment]
I am glad it was not me going mad 👍

**Comment by freddydk** [comment]
Thanks a lot for your help - this could have caused a lot of partners some hard times (maybe already does).

**Comment by freddydk** [comment]
If you are using 10.0.14393.3808 still, then the generic image should be available in approx. 15 minutes

**Comment by darrenwaller** [comment]
Brilliant, thanks for sorting so quickly

**Comment by darrenwaller** [comment]
That seems to be working, thank you Freddy

**Comment by freddydk** [comment]
Shipped in generic image 0.1.0.13
